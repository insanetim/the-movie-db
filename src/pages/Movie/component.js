import { Row, Col, Typography, Tag, Popover } from 'antd'
import Icon, { HeartOutlined, HeartFilled, BookOutlined, BookFilled, PlusCircleOutlined } from '@ant-design/icons'
import { format } from 'date-fns'
import ISO6391 from 'iso-639-1'
import { take, toUpper } from 'ramda'

import CreditsItem from 'src/components/Movie/CreditsItem'
import ImageGallery from 'src/components/Movie/ImageGallery'
import PopoverContent from 'src/components/Movie/PopoverContent'
import Loading from 'src/components/Loading'
import Error from 'src/components/Error'
import { convertDuration, convertMoney } from 'src/utils'
import useContainer from './hook'

const Movie = () => {
  const { movie, loading, error, handleFavoriteClick, handleWatchlistClick, popoverOpen, setPopoverOpen } =
    useContainer()

  if (loading) {
    return (
      <div className='top-margin'>
        <Loading />
      </div>
    )
  }

  if (error) {
    return (
      <div className='top-margin'>
        <Error error={error} />
      </div>
    )
  }

  return (
    <>
      {movie.images.length > 0 && <ImageGallery images={movie.images} />}
      <div className='top-margin'>
        <Row justify='center'>
          <Col span={20}>
            <Typography.Title>
              <span>
                {movie.title}
                {movie.release_date && ` (${format(new Date(movie.release_date), 'yyyy')})`}
              </span>{' '}
              <Popover
                title='Add movie to list'
                trigger='click'
                placement='top'
                destroyTooltipOnHide
                open={popoverOpen}
                onOpenChange={open => setPopoverOpen(open)}
                content={
                  <PopoverContent
                    movieId={movie.id}
                    setPopoverOpen={setPopoverOpen}
                  />
                }
              >
                <PlusCircleOutlined data-testid='addMovieToListPopover' />
              </Popover>{' '}
              <Icon
                component={movie.accountStates.favorite ? HeartFilled : HeartOutlined}
                onClick={handleFavoriteClick}
              />{' '}
              <Icon
                component={movie.accountStates.watchlist ? BookFilled : BookOutlined}
                onClick={handleWatchlistClick}
              />
            </Typography.Title>
            {movie.overview && (
              <>
                <Typography.Title level={3}>Overview</Typography.Title>
                <Typography.Paragraph>{movie.overview}</Typography.Paragraph>
              </>
            )}
          </Col>
        </Row>
        <Row justify='center'>
          <Col span={20}>
            <Typography.Paragraph>
              <b>Original Language: </b>
              <span>{ISO6391.getName(movie.original_language)}</span>
            </Typography.Paragraph>
          </Col>
          {movie.runtime !== 0 && (
            <Col span={20}>
              <Typography.Paragraph>
                <b>Runtime: </b>
                <span>{convertDuration(movie.runtime)}</span>
              </Typography.Paragraph>
            </Col>
          )}
          {movie.budget !== 0 && (
            <Col span={20}>
              <Typography.Paragraph>
                <b>Budget: </b>
                <span>{convertMoney(movie.budget)}</span>
              </Typography.Paragraph>
            </Col>
          )}
          {movie.revenue !== 0 && (
            <Col span={20}>
              <Typography.Paragraph>
                <b>Revenue: </b>
                <span>{convertMoney(movie.revenue)}</span>
              </Typography.Paragraph>
            </Col>
          )}
          {movie.genres.length > 0 && (
            <Col span={20}>
              <Typography.Paragraph>
                <b>Genres: </b>
                {movie.genres.map(({ id, name }) => (
                  <Tag key={id}>{toUpper(name)}</Tag>
                ))}
              </Typography.Paragraph>
            </Col>
          )}
        </Row>
        {movie.credits.cast.length > 0 && (
          <>
            <Row>
              <Col
                className='top-margin'
                span={10}
                offset={2}
              >
                <Typography.Title level={3}>Casts</Typography.Title>
              </Col>
            </Row>
            <Row
              gutter={8}
              justify='center'
            >
              <Col span={20}>
                <Row
                  gutter={{
                    xs: 8,
                    sm: 16,
                    md: 24,
                    lg: 32
                  }}
                >
                  {take(12, movie.credits.cast).map(item => (
                    <CreditsItem
                      key={item.credit_id}
                      profilePath={item.profile_path}
                      title={item.name}
                      description={item.character}
                    />
                  ))}
                </Row>
              </Col>
            </Row>
          </>
        )}
        {movie.credits.crew.length > 0 && (
          <>
            <Row>
              <Col
                className='top-margin'
                span={10}
                offset={2}
              >
                <Typography.Title level={3}>Crew</Typography.Title>
              </Col>
            </Row>
            <Row
              gutter={8}
              justify='center'
            >
              <Col span={20}>
                <Row
                  gutter={{
                    xs: 8,
                    sm: 16,
                    md: 24,
                    lg: 32
                  }}
                >
                  {take(12, movie.credits.crew).map(item => (
                    <CreditsItem
                      key={item.credit_id}
                      profilePath={item.profile_path}
                      title={item.name}
                      description={item.job}
                    />
                  ))}
                </Row>
              </Col>
            </Row>
          </>
        )}
      </div>
    </>
  )
}

export default Movie
