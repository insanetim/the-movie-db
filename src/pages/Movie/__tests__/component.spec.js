import { shallow } from 'enzyme'
import { mergeDeepRight } from 'ramda'

import Movie from '../component'
import useContainer from '../hook'

const mockedHookData = {
  movie: {
    id: 1,
    title: 'test/title',
    release_date: new Date(),
    overview: 'test/overview',
    original_language: 'EN',
    runtime: 90,
    budget: 1000000,
    revenue: 1500000,
    genres: [
      { id: 1, name: 'action' },
      { id: 2, name: 'horror' }
    ],
    images: [{ file_path: 'test/image' }],
    credits: {
      cast: [{ credit_id: 1, profile_path: 'test/image', name: 'John Doe', character: 'Neo' }],
      crew: [{ credit_id: 1, profile_path: 'test/image', name: 'John Doe', job: 'Director' }]
    },
    accountStates: {
      favorite: false,
      watchlist: false
    }
  },
  loading: false,
  error: null,
  handleFavoriteClick: jest.fn(),
  handleWatchlistClick: jest.fn(),
  popoverOpen: false,
  setPopoverOpen: jest.fn()
}
jest.mock('../hook', () => jest.fn(() => mockedHookData))

describe('Movie component tests', () => {
  let component = shallow(<Movie />)

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot()
  })

  it('matches snapshot with other data', () => {
    useContainer.mockReturnValueOnce(
      mergeDeepRight(mockedHookData, {
        movie: {
          accountStates: {
            favorite: true,
            watchlist: true
          }
        },
        popoverOpen: true
      })
    )
    component = shallow(<Movie />)

    expect(component).toMatchSnapshot()
  })

  it('Popover matches snapshot', () => {
    component = shallow(<Movie />)

    expect(component.find('Popover').renderProp('onOpenChange')()).toMatchSnapshot()
  })

  it('matches snapshot with loading', () => {
    mockedHookData.loading = true
    component = shallow(<Movie />)

    expect(component).toMatchSnapshot()
  })

  it('matches snapshot with error', () => {
    mockedHookData.loading = false
    mockedHookData.error = { message: 'test/error' }
    component = shallow(<Movie />)

    expect(component).toMatchSnapshot()
  })
})
