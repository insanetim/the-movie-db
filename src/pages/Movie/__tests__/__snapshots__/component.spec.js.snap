// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie component tests Popover matches snapshot 1`] = `""`;

exports[`Movie component tests matches snapshot 1`] = `
<Fragment>
  <ImageGallery
    images={
      [
        {
          "file_path": "test/image",
        },
      ]
    }
  />
  <div
    className="top-margin"
  >
    <Row
      justify="center"
    >
      <Col
        span={20}
      >
        <ForwardRef>
          <span>
            test/title
             (2023)
          </span>
           
          <Popover
            content={
              <PopoverContent
                movieId={1}
                setPopoverOpen={[MockFunction]}
              />
            }
            destroyTooltipOnHide={true}
            onOpenChange={[Function]}
            open={false}
            placement="top"
            title="Add movie to list"
            trigger="click"
          >
            <ForwardRef(PlusCircleOutlined) />
          </Popover>
           
          <AntdIcon
            component={
              {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              }
            }
            onClick={[MockFunction]}
          />
           
          <AntdIcon
            component={
              {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              }
            }
            onClick={[MockFunction]}
          />
        </ForwardRef>
        <ForwardRef
          level={3}
        >
          Overview
        </ForwardRef>
        <ForwardRef>
          test/overview
        </ForwardRef>
      </Col>
    </Row>
    <Row
      justify="center"
    >
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Original Language: 
          </b>
          <span />
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Runtime: 
          </b>
          <span>
            1h 30m
          </span>
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Budget: 
          </b>
          <span>
            $1,000,000.00
          </span>
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Revenue: 
          </b>
          <span>
            $1,500,000.00
          </span>
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Genres: 
          </b>
          <Tag
            key="1"
          >
            ACTION
          </Tag>
          <Tag
            key="2"
          >
            HORROR
          </Tag>
        </ForwardRef>
      </Col>
    </Row>
    <Row>
      <Col
        className="top-margin"
        offset={2}
        span={10}
      >
        <ForwardRef
          level={3}
        >
          Casts
        </ForwardRef>
      </Col>
    </Row>
    <Row
      gutter={8}
      justify="center"
    >
      <Col
        span={20}
      >
        <Row
          gutter={
            {
              "lg": 32,
              "md": 24,
              "sm": 16,
              "xs": 8,
            }
          }
        >
          <CreditsItem
            description="Neo"
            key="1"
            profilePath="test/image"
            title="John Doe"
          />
        </Row>
      </Col>
    </Row>
    <Row>
      <Col
        className="top-margin"
        offset={2}
        span={10}
      >
        <ForwardRef
          level={3}
        >
          Crew
        </ForwardRef>
      </Col>
    </Row>
    <Row
      gutter={8}
      justify="center"
    >
      <Col
        span={20}
      >
        <Row
          gutter={
            {
              "lg": 32,
              "md": 24,
              "sm": 16,
              "xs": 8,
            }
          }
        >
          <CreditsItem
            description="Director"
            key="1"
            profilePath="test/image"
            title="John Doe"
          />
        </Row>
      </Col>
    </Row>
  </div>
</Fragment>
`;

exports[`Movie component tests matches snapshot with error 1`] = `
<div
  className="top-margin"
>
  <Error
    error={
      {
        "message": "test/error",
      }
    }
  />
</div>
`;

exports[`Movie component tests matches snapshot with loading 1`] = `
<div
  className="top-margin"
>
  <Loading />
</div>
`;

exports[`Movie component tests matches snapshot with other data 1`] = `
<Fragment>
  <ImageGallery
    images={
      [
        {
          "file_path": "test/image",
        },
      ]
    }
  />
  <div
    className="top-margin"
  >
    <Row
      justify="center"
    >
      <Col
        span={20}
      >
        <ForwardRef>
          <span>
            test/title
             (2023)
          </span>
           
          <Popover
            content={
              <PopoverContent
                movieId={1}
                setPopoverOpen={[MockFunction]}
              />
            }
            destroyTooltipOnHide={true}
            onOpenChange={[Function]}
            open={true}
            placement="top"
            title="Add movie to list"
            trigger="click"
          >
            <ForwardRef(PlusCircleOutlined) />
          </Popover>
           
          <AntdIcon
            component={
              {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              }
            }
            onClick={[MockFunction]}
          />
           
          <AntdIcon
            component={
              {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              }
            }
            onClick={[MockFunction]}
          />
        </ForwardRef>
        <ForwardRef
          level={3}
        >
          Overview
        </ForwardRef>
        <ForwardRef>
          test/overview
        </ForwardRef>
      </Col>
    </Row>
    <Row
      justify="center"
    >
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Original Language: 
          </b>
          <span />
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Runtime: 
          </b>
          <span>
            1h 30m
          </span>
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Budget: 
          </b>
          <span>
            $1,000,000.00
          </span>
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Revenue: 
          </b>
          <span>
            $1,500,000.00
          </span>
        </ForwardRef>
      </Col>
      <Col
        span={20}
      >
        <ForwardRef>
          <b>
            Genres: 
          </b>
          <Tag
            key="1"
          >
            ACTION
          </Tag>
          <Tag
            key="2"
          >
            HORROR
          </Tag>
        </ForwardRef>
      </Col>
    </Row>
    <Row>
      <Col
        className="top-margin"
        offset={2}
        span={10}
      >
        <ForwardRef
          level={3}
        >
          Casts
        </ForwardRef>
      </Col>
    </Row>
    <Row
      gutter={8}
      justify="center"
    >
      <Col
        span={20}
      >
        <Row
          gutter={
            {
              "lg": 32,
              "md": 24,
              "sm": 16,
              "xs": 8,
            }
          }
        >
          <CreditsItem
            description="Neo"
            key="1"
            profilePath="test/image"
            title="John Doe"
          />
        </Row>
      </Col>
    </Row>
    <Row>
      <Col
        className="top-margin"
        offset={2}
        span={10}
      >
        <ForwardRef
          level={3}
        >
          Crew
        </ForwardRef>
      </Col>
    </Row>
    <Row
      gutter={8}
      justify="center"
    >
      <Col
        span={20}
      >
        <Row
          gutter={
            {
              "lg": 32,
              "md": 24,
              "sm": 16,
              "xs": 8,
            }
          }
        >
          <CreditsItem
            description="Director"
            key="1"
            profilePath="test/image"
            title="John Doe"
          />
        </Row>
      </Col>
    </Row>
  </div>
</Fragment>
`;
